File: Binary-Search/occurrence-of-element-in-a-sorted-array.py
# Time  : O(logn)
# Space : O(1)

def count(A: [int], n: int, target: int) -> int:
    lo, hi = 0, n - 1
    first = -1

    while lo <= hi:
        mid = lo + (hi - lo) // 2
        if A[mid] < target:
            lo = mid + 1
        elif A[mid] > target:
            hi = mid - 1
        else:
            first = mid
            hi = mid - 1

    lo, hi = 0, n - 1
    last = -1

    while lo <= hi:
        mid = lo + (hi - lo) // 2
        if A[mid] < target:
            lo = mid + 1
        elif A[mid] > target:
            hi = mid - 1
        else:
            last = mid
            lo = mid + 1

    if first == -1:
        return 0
    return last - first + 1

File: Binary-Search/occurrence-of-element-in-a-sorted-array.cpp
// Time  : O(logn)
// Space : O(1)
class Solution{
public:
	int count(int A[], int n, int x) {
    int lo = 0;
    int hi = n;
    while (lo < hi) {
      int mid = (lo + hi)>>1;
      if (A[mid] < x) lo = mid+1;
      else hi = mid;
    }
    if (lo == n || A[lo] != x) return 0;
    int first = lo;
    lo = 0;
    hi = n;
    while (lo < hi) {
      int mid = (lo + hi)>>1;
      if (A[mid] <= x) lo = mid+1;
      else hi = mid;
    }
    int last = lo-1;
    return last - first + 1;
	}
};